### PR TITLE
Implement WSD without cooldown

### PR DESCRIPTION
**Pull request by mikasenghaas**
Refactors the setup of LR schedulers a little bit with the main intention to support WSD without cooldown with infinite training. Also fixes a reporting issue where we call `scheduler.step()` before logging the current learning rate (i.e. we report the LR one step off). Also fixes a bug from before when `max_steps == warmup_steps`.

Check the schedules on [W&B](https://wandb.ai/primeintellect/mika/workspace?nw=0rkxwwyka9es).

Screenshot 2025-09-22 at 11 38 50 PM

---

**GitHub Issue**: #968, #985
**Linear Issue**: Resolves PRIMERL-121, PRIMERL-129

### CODE REVIEW

**Review by samsja — APPROVED** [review]
lgtm
